gcc/rust/ast/rust-ast-dump.cc

```diff
@@ -1126,22 +1130,138 @@ Dump::visit (ExternBlock &block)
   stream << "\n" << indentation << "}\n";
 }
 
-// rust-macro.h
+static std::pair<char, char>
+get_delimiters (DelimType delim)
+{
+  auto start_delim = '\0';
+  auto end_delim = '\0';
+
+  switch (delim)
+    {
+    case PARENS:
+      start_delim = '(';
+      end_delim = ')';
+      break;
+    case SQUARE:
+      start_delim = '[';
+      end_delim = ']';
+      break;
+    case CURLY:
+      start_delim = '{';
+      end_delim = '}';
+      break;
+    default:
+      gcc_unreachable ();
+    }
+
+  return {start_delim, end_delim};
+}
+
 void
 Dump::visit (MacroMatchFragment &match)
-{}
+{
+  stream << match.get_ident () << ':' << match.get_frag_spec ().as_string ();
+}
 
 void
-Dump::visit (MacroMatchRepetition &match)
-{}
+Dump::visit (MacroMatchRepetition &repetition)
+{
+  stream << "$(";
+
+  for (auto &match : repetition.get_matches ())
+    {
+      match->accept_vis (*this);
+      stream << ' ';
+    }
+
+  auto op_char = '\0';
+  switch (repetition.get_op ())
+    {
+    case MacroMatchRepetition::ANY:
+      op_char = '*';
+      break;
+    case MacroMatchRepetition::ONE_OR_MORE:
+      op_char = '+';
+      break;
+    case MacroMatchRepetition::ZERO_OR_ONE:
+      op_char = '?';
+      break;
+    case MacroMatchRepetition::NONE:
+      break;
+    }
+
+  stream << ')';
+
+  if (repetition.has_sep ())
+    stream << repetition.get_sep ()->as_string ();
+
+  stream << op_char;
+}
 
 void
 Dump::visit (MacroMatcher &matcher)
-{}
+{
+  auto delimiters = get_delimiters (matcher.get_delim_type ());
+
+  stream << delimiters.first;
+
+  for (auto &match : matcher.get_matches ())
+    {
+      match->accept_vis (*this);
+      stream << ' ';
+    }
+
+  stream << delimiters.second;
+}
+
+void
+Dump::emit_transcriber (const MacroTranscriber &transcriber)
+{
+  emit_token_tree (transcriber.get_token_tree ());
+}
+
+void
+Dump::emit_token_tree (const DelimTokenTree &tree)
+{
+  auto tokens = tree.to_token_stream ();
+
+  indentation.increment ();
+  stream << '\n' << indentation;
```

Comment:
This \n should not be here. See
```rust
macro_rules! foo {
        (l:tt ) => // HERE SHOULD NOT BE A NEW NEW LINE
                 { bar ! ( $ l ) ; }
        ;
}
```

Reply:
The issue with this is that the delimiters are part of the token stream. We could change it and avoid printing the delimiters (something like `for (size_t i = 1; i < tokens.length() - 1; i++)` etc) but that could be fixed in a later PR imo